### ISSUE TITLE
"intrinsic" on test eax, eax

### ISSUE THREAD

**Issue by mrexodia** [yegord/snowman#2]
Hi,

Recently I implemented snowman in x64dbg and we noticed a weird thing. This code:

```
00007FF66E0D1444 | 48 83 EC 38              | sub rsp,38                              |
00007FF66E0D1448 | 48 83 64 24 20 00        | and qword ptr ss:[rsp+20],0             |
00007FF66E0D144E | 41 B9 01 00 00 00        | mov r9d,1                               |
00007FF66E0D1454 | 4C 8D 44 24 40           | lea r8,qword ptr ss:[rsp+40]            |
00007FF66E0D1459 | 41 8D 51 10              | lea edx,dword ptr ds:[r9+10]            |
00007FF66E0D145D | 48 C7 C1 FE FF FF FF     | mov rcx,FFFFFFFFFFFFFFFE                |
00007FF66E0D1464 | E8 AB FA FC FF           | call 7FF66E0A0F14                       |
00007FF66E0D1469 | 85 C0                    | test eax,eax                            |
00007FF66E0D146B | 78 0A                    | js plzplz.7FF66E0D1477                  |
00007FF66E0D146D | 80 7C 24 40 00           | cmp byte ptr ss:[rsp+40],0              |
00007FF66E0D1472 | 75 03                    | jnz plzplz.7FF66E0D1477                 |
00007FF66E0D1474 | CC                       | int3                                    |
00007FF66E0D1475 | EB 00                    | jmp plzplz.7FF66E0D1477                 |
00007FF66E0D1477 | 48 83 C4 38              | add rsp,38                              |
00007FF66E0D147B | C3                       | ret                                     |
```

Shows as:

```
void fun_140001444() {
    int32_t eax1;
    signed char v2;

    eax1 = fun_13ffd0f14();
    if (!"intrinsic"() && v2 == 0) {
    }
    return;
}
```

The tree appears to go in infinity recursion: ![tree](http://i.imgur.com/lPJcTb1.png)

Here is a binary (modified to only have this function, no malware): https://mega.co.nz/#!TgZz2LJa!RhX6Cc-SUiw8-IQufGamfLncm7PorI5odyCQf8mkk7Y

**Comment by yegord** [comment]
If the question is, where the intrinsic comes from, then the answer is — from the test instruction, see X86InstructionAnalyzer.cpp:1063, `sf ^= intrinsic()`. Perhaps, one can write a special expression for the case when the first operand is equal to the second operand. Feel free to send a patch. Or I will fix it eventually myself.

The infiniteness of the tree is a feature: a term has a pointer to the parent statement, a statement has a pointer to its terms. :)

**Comment by mrexodia** [comment]
My decompiled code would be:

```
void fun_140001444() {
    int32_t eax1;
    signed char v2;

    eax1 = fun_13ffd0f14();
    if (eax1 <= -1 && v2 == 0) {
        __asm("int3");
    }
    return;
}
```

Since `test eax, eax` will set the sign flag to 1 if the result is smaller or equal -1. The flag is determined by the most significant bit, which will be set if the number is negative (seen from a signed perspective). A number AND itself is obviously the number again, so this is just a test to see if the result was negative.

The `int3` instruction is also removed from the output for some reason.

**Comment by mrexodia** [comment]
In a more generic way, `test op1, op2` and a `js` after that would become something like `if( (op1 & op2) <= -1)`

**Comment by mrexodia** [comment]
I came up with this code, but I have absolutely no clue what I'm doing :+1: 

```
            case UD_Itest: {
                _[
                    cf ^= constant(0),
                    pf ^= intrinsic()
                ];

                if (operandsAreTheSame(0, 1)) {
                    _[zf ^= operand(0) == constant(0)];
                    _[sf ^= signed_(operand(0)) < constant(0)];
                } else {
                    _[zf ^= (operand(0) & operand(1)) == constant(0)];
                    _[sf ^= signed_(operand(0) & operand(1)) < constant(0)];
                }

                _[
                    of ^= constant(0),
                    af ^= undefined(),
                    kill(pseudo_flags)
                ];
                break;
            }
```

**Comment by yegord** [comment]
The code looks correct. Do you want to send a pull request, or I just commit this myself?

**Comment by mrexodia** [comment]
I'll do a pull later on. I also fixed something else.

**Comment by yegord** [comment]
Resolved with a22a076c859a08542eb812c7ff5cb264aa3dc72c.
